### ISSUE TITLE
[qt] qt/6.5.0: broken recipe

### ISSUE THREAD

**Issue by ethindp** [conan-io/conan-center-index#17449]
### Description

I'm trying to use Qt in a project but get this error when I run conan install:

```
======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=20
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=20
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows


======== Computing dependency graph ========
Graph root
    conanfile.py (mud/0.1.0): c:\Users\ethin\source\mudserver\conanfile.py
Requirements
    boost/1.81.0#9bd7ae86a881631b6cc76590621e470b - Cache
    bzip2/1.0.8#411fc05e80d47a89045edc1ee6f23c1d - Cache
    zlib/1.2.13#e377bee636333ae348d51ca90874e353 - Cache
Build requirements
    b2/4.9.6#2e02945ad41739b8051247d14fe8297d - Cache
Resolved version ranges
    boost/[>=1.81.0]: boost/1.81.0
*********************************************************
Recipe 'qt/6.4.2' seems broken.
It is possible that this recipe is not Conan 2.0 ready
If the recipe comes from ConanCenter check: https://conan.io/cci-v2.html
If it is your recipe, check if it is updated to 2.0
*********************************************************

ERROR: Package 'qt/6.4.2' not resolved: qt/6.4.2: Cannot load recipe.
Error loading conanfile at 'C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py': Unable to load conanfile in C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py", line 18, in <module>
    from conans import RunEnvironment, tools
ImportError: cannot import name 'RunEnvironment'
.
```

Moving to an older version doesn't solve the problem.

### Package and Environment Details

* Package Name/Version: **qt/6.5.0**
* Operating System+version: **Windows 11 21H2 (AMD64) build 22000.1817**
* Compiler+version: **MSVC 19.34.31944**
* Docker image: **none**
* Conan version: **conan 2.0.4**
* Python version: **Python 3.10.9**

### Conan profile

Host profile:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=20
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows

Build profile:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=20
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows

### Steps to reproduce

conan install . --output-folder=build --build=missing

### Logs

<details><summary>Click to expand log</summary>

```

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=20
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=20
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows


======== Computing dependency graph ========
Graph root
    conanfile.py (mud/0.1.0): c:\Users\ethin\source\mudserver\conanfile.py
Requirements
    boost/1.81.0#9bd7ae86a881631b6cc76590621e470b - Cache
    bzip2/1.0.8#411fc05e80d47a89045edc1ee6f23c1d - Cache
    zlib/1.2.13#e377bee636333ae348d51ca90874e353 - Cache
Build requirements
    b2/4.9.6#2e02945ad41739b8051247d14fe8297d - Cache
Resolved version ranges
    boost/[>=1.81.0]: boost/1.81.0
Traceback (most recent call last):
  File "conans\client\loader.py", line 314, in load_python_file
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py", line 18, in <module>
    from conans import RunEnvironment, tools
ImportError: cannot import name 'RunEnvironment'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "conans\client\loader.py", line 55, in load_basic_module
  File "conans\client\loader.py", line 288, in parse_conanfile
  File "conans\client\loader.py", line 352, in load_python_file
conans.errors.ConanException: Unable to load conanfile in C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py", line 18, in <module>
    from conans import RunEnvironment, tools
ImportError: cannot import name 'RunEnvironment'


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "conans\client\loader.py", line 184, in load_conanfile
  File "conans\client\loader.py", line 79, in load_basic_module
conans.errors.ConanException: Error loading conanfile at 'C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py': Unable to load conanfile in C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py", line 18, in <module>
    from conans import RunEnvironment, tools
ImportError: cannot import name 'RunEnvironment'


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "conans\client\graph\graph_builder.py", line 246, in _create_new_node
  File "conans\client\graph\graph_builder.py", line 213, in _resolve_recipe
  File "conans\client\loader.py", line 186, in load_conanfile
conans.errors.ConanException: qt/6.4.2: Cannot load recipe.
Error loading conanfile at 'C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py': Unable to load conanfile in C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py", line 18, in <module>
    from conans import RunEnvironment, tools
ImportError: cannot import name 'RunEnvironment'


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "conan\cli\cli.py", line 171, in run
  File "conan\cli\command.py", line 157, in run
  File "conan\cli\commands\install.py", line 76, in install
  File "conans\client\graph\graph.py", line 300, in report_graph_error
  File "conans\client\graph\graph_builder.py", line 51, in load_graph
  File "conans\client\graph\graph_builder.py", line 89, in _expand_require
  File "conans\client\graph\graph_builder.py", line 248, in _create_new_node
conans.client.graph.graph_error.GraphMissingError: Package 'qt/6.4.2' not resolved: qt/6.4.2: Cannot load recipe.
Error loading conanfile at 'C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py': Unable to load conanfile in C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py", line 18, in <module>
    from conans import RunEnvironment, tools
ImportError: cannot import name 'RunEnvironment'
.

*********************************************************
Recipe 'qt/6.4.2' seems broken.
It is possible that this recipe is not Conan 2.0 ready
If the recipe comes from ConanCenter check: https://conan.io/cci-v2.html
If it is your recipe, check if it is updated to 2.0
*********************************************************

ERROR: Package 'qt/6.4.2' not resolved: qt/6.4.2: Cannot load recipe.
Error loading conanfile at 'C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py': Unable to load conanfile in C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "C:\Users\ethin\.conan2\p\qt77d2605fccfa8\e\conanfile.py", line 18, in <module>
    from conans import RunEnvironment, tools
ImportError: cannot import name 'RunEnvironment'
.

```

(This happens with 6.5.0 too, me moving to 6.4.2 was an attempt to solve it.)
</details>

**Comment by sabelka** [comment]
The recipe for qt6 has not yet been ported to conan v2. But there are ongoing efforts in this direction in #17539.

**Comment by MartinDelille** [comment]
Review would help this PR to be merged!

**Comment by nrbnlulu** [comment]
#17539 Didn't fixed v2 yet... This is what I get
### Environment details

* Operating System+version:ubuntu 22.0.4
* Compiler+version:
```console
Ubuntu clang version 14.0.0-1ubuntu1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```
* Conan version: 2.0.6
* Python version: 3.10.6


### Steps to reproduce

This is the conan file I used
```py
from conan import ConanFile
from conan.tools.cmake import CMakeToolchain, CMake, cmake_layout, CMakeDeps


class CompressorRecipe(ConanFile):
    settings = "os", "compiler", "build_type", "arch"

    def requirements(self):
        self.requires("qt/6.5.0")

    def layout(self) -> None:
        cmake_layout(self)
    def generate(self):
        deps = CMakeDeps(self)
        deps.generate()
        tc = CMakeToolchain(self)
        tc.generate()

    def build(self):
        cmake = CMake(self)
        cmake.configure()
```
run `conan lock create .`

### Logs

```console
======== Computing dependency graph ========
Graph root
    conanfile.py: /home/nir/Desktop/opensource/my/qtgql/examples/qtify/qtify-client/conanfile.py
*********************************************************
Recipe 'qt/6.5.0' seems broken.
It is possible that this recipe is not Conan 2.0 ready
If the recipe comes from ConanCenter check: https://conan.io/cci-v2.html
If it is your recipe, check if it is updated to 2.0
*********************************************************

ERROR: Package 'qt/6.5.0' not resolved: qt/6.5.0: Cannot load recipe.
Error loading conanfile at '/home/nir/.conan2/p/qta572734c3567f/e/conanfile.py': Unable to load conanfile in /home/nir/.conan2/p/qta572734c3567f/e/conanfile.py
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/nir/.conan2/p/qta572734c3567f/e/conanfile.py", line 18, in <module>
    from conans import RunEnvironment, tools
ImportError: cannot import name 'RunEnvironment' from 'conans' (/home/nir/Desktop/opensource/my/qtgql/examples/qtify/.venv/lib/python3.10/site-packages/conans/__init__.py) 
```

**Comment by Marwan-imverse** [comment]
using latest master, I get the following error:```

qt/6.4.2: Error in source() method, line 598
        if self.settings.os == "Windows":
        ConanException: 'self.settings' access in 'source()' method is forbidden
```

**Comment by MartinDelille** [comment]
I wonder if it should not be `self.info.settings.os` there.

**Comment by Marwan-imverse** [comment]
TBH I really really don't have time to work on the fix. Just laying things out to whoever will do this (hopefully 🤞 )

**Comment by EstebanDugueperoux2** [comment]
Hi,

I get same error today, and it works by using **self.info.settings.os** instead, but it means there is an error in doc https://docs.conan.io/2/reference/conanfile/methods/source.html#source-caching .
I have created a pull request.

**Comment by MartinDelille** [comment]
You might try to checkout if the fix in this PR solve your issue: https://github.com/conan-io/conan-center-index/pull/17942

**Comment by Marwan-imverse** [comment]
That PR is very outdated. It stops at version 6.4.1 and actually uses `self.copy()` which is not Conan 2 compatible.

**Comment by MartinDelille** [comment]
Neither the [conanfile.py](https://github.com/conan-io/conan-center-index/blob/8a0cbd3e7d11570772832e485c13fa7a4cc901f3/recipes/qt/6.x.x/conanfile.py) nor the [test_package/conanfile.py](https://github.com/conan-io/conan-center-index/blob/8a0cbd3e7d11570772832e485c13fa7a4cc901f3/recipes/qt/6.x.x/test_package/conanfile.py) contains the word copy.

**Comment by MartinDelille** [comment]
You should try the following:

```
gh clone https://github.com/conan-io/conan-center-index
cd conan-center-index/recipes
gh pr checkout 17942
conan export qt/6.x.x qt/6.5.0
```

**Comment by Marwan-imverse** [comment]
I cloned the branch `https://github.com/ericLemanissier/conan-center-index.git`
I'll try just checking out the PR then

**Comment by Marwan-imverse** [comment]
Still not working:
```
ERROR: qt/6.5.1@imverse/externals: Error in source() method, line 605
        apply_conandata_patches(self)
        ConanException: Failed to apply patch: C:\Users\Marwan\.conan2\p\qt45970e9e8d8ec\s\patches/qt6.5.0-pri-helpers-fix.diff
```

**Comment by ericLemanissier** [comment]
@Marwan-imverse  https://github.com/conan-io/conan-center-index/pull/17942 is now merged. you can simply retry the conan command you were calling with `-u` to get the last revision of the recipe

**Comment by Marwan-imverse** [comment]
ah good, I'll pull the latest master then

**Comment by Marwan-imverse** [comment]
Getting this error with Qt 6.5.1:
```
fatal error C1083: Cannot open include file: '../../../../../../../../../../qt9fdb84fb6e264/s/src/qtdeclarative/src/quickdialogs/quickdialogsquickimpl/qquickabstractcolorpicker_p.h': No such file or directory
```

**Comment by Marwan-imverse** [comment]
If it helps, I'm building with the following options:
```
-o qtdeclarative=True -o qtimageformats=True -o qtlanguageserver=True -o qtshadertools=True -o qtsvg=True
```

**Comment by MartinDelille** [comment]
Cool! If you are using qml, you might run into the issue that I tried to fixed here without success:

https://github.com/conan-io/conan-center-index/pull/17925

**Comment by ericLemanissier** [comment]
Please provide the complete command you invoke, and the full console log.
This patching error does not happen in conan-center CI, https://c3i.jfrog.io/artifactory/misc/logs/pr/17942/33-windows-visual_studio/qt/6.4.2/0d56a7497637304e44735904ac26294e5a1d4fa7-build.txt so you probably want to retry with a clean conan local cache, and make sure you are actually using the recipe without any modification from CCI : https://github.com/conan-io/conan-center-index/tree/master/recipes/qt/6.x.x

**Comment by Marwan-imverse** [comment]
I found this concerning the error I get for Qt 6.5.12 maybe it can help? https://github.com/microsoft/vcpkg/issues/30446

**Comment by Marwan-imverse** [comment]
> Please provide the complete command you invoke, and the full console log. This patching error does not happen in conan-center CI, https://c3i.jfrog.io/artifactory/misc/logs/pr/17942/33-windows-visual_studio/qt/6.4.2/0d56a7497637304e44735904ac26294e5a1d4fa7-build.txt so you probably want to retry with a clean conan local cache, and make sure you are actually using the recipe without any modification from CCI : https://github.com/conan-io/conan-center-index/tree/master/recipes/qt/6.x.x

I am actually using the recipe to build the licensed sources of Qt. They apparently differ slightly from the LGPL sources. I am going to try and apply fuzzy patching when I have some time. I had this issue with other patches that were running fine with Conan 1.X. Conan 2 is much stricter with applying patches and it fails if the line numbers are not exactly the same between the two files.

**Comment by Marwan-imverse** [comment]
The "good" news is that after switching to fuzzy patching, I solved [the error above](https://github.com/conan-io/conan-center-index/issues/17449#issuecomment-1640347476).

FYI I replaced line 606: `apply_conandata_patches(self)` with:
```python
for it in self.conan_data.get("patches", {}).get(self.version, []):
            patch(self, **it, fuzz=True)
```

I actually opened a [feature request](https://github.com/conan-io/conan/issues/14355) to add the fuzzy patching option to `apply_conandata_patches()`. 

However, I now have the exact same error I have with qt 6.5.1.

Full command invoked:
`conan create . --version=6.5.1 --user imverse --channel externals -pr:h windows_VS17 -pr:b windows_VS17 -o qtdeclarative=True -o qtimageformats=True -o qtlanguageserver=True -o qtshadertools=True -o qtsvg=True`

And here is the full output: [qt_conan.log](https://github.com/conan-io/conan-center-index/files/12142882/qt_conan.log)

**Comment by ericLemanissier** [comment]
Ok:

- does the file `C:\Users\Marwan\.conan2\p\b\qtbd9c363298b7a\b\build\qtbase\include\QtQuickDialogs2QuickImpl\6.4.2\QtQuickDialogs2QuickImpl\private/qquickabstractcolorpicker_p.h`  exist ?
- does the same error happen if you use conan-center recipe without modification ?
- Does your recipe have the exact same `source(self)` method as conan-center, especially https://github.com/conan-io/conan-center-index/blob/master/recipes/qt/6.x.x/conanfile.py#L600 ?
- Does the error also happens if you use a shorter path to conan local cache (eg `c:\conan2` instead of `C:\Users\Marwan\.conan2` ?

**Comment by Marwan-imverse** [comment]
>     * does the file `C:\Users\Marwan\.conan2\p\b\qtbd9c363298b7a\b\build\qtbase\include\QtQuickDialogs2QuickImpl\6.4.2\QtQuickDialogs2QuickImpl\private/qquickabstractcolorpicker_p.h`  exist ?
Yes it exists, the file that includes it exists, and the relative path from the including file to the included file exists.

>     * does the same error happen if you use conan-center recipe without modification ?
Yes, using vanilla recipe from conancenter main branch.

>     * Does your recipe have the exact same `source(self)` method as conan-center, especially https://github.com/conan-io/conan-center-index/blob/master/recipes/qt/6.x.x/conanfile.py#L600 ?
No, I copy the cources from the folder containing the licensed source files. But I do have the error with vanilla recipe from conancenter as explained in my previous answer.

>     * Does the error also happens if you use a shorter path to conan local cache (eg `c:\conan2` instead of `C:\Users\Marwan\.conan2` ?
That did the trick! I set the cache folder to `C:/C2` and both vanilla and my recipe work!

**Comment by nrbnlulu** [comment]
@Marwan-imverse 
Did you had a problme with QML plugins not being created?

**Comment by nrbnlulu** [comment]
I am willing to fix this issue for v2. Can someone give me some guidelines? CC/ @MartinDelille @ericLemanissier

**Comment by MartinDelille** [comment]
@nrbnlulu Which issue ? Isn't qt ready for v2 yet ?

**Comment by nrbnlulu** [comment]
1. The workflows for v2 are not passing.
2. It is not in the v2 ready list
3. I couldn't get qt quick to work. Somw plugins are not beeing built.

**Comment by ericLemanissier** [comment]
> * The workflows for v2 are not passing.
The current failure is in the test package, on macos when qt is shared, because double-conversion dynamic library cannot be found (because of SIP). see https://c3i.jfrog.io/c3i/misc-v2/logs/pr/18697/8-macos-clang/qt/6.5.1//041266c9d3ce1fdb080c1c13b2dbff65d8f3e425-test.txt. Don't hesitate to work on this. In the mean-time, I'll disable the shared version of qt on macos, to unblock CI.
> * It is not in the v2 ready list
Yes it is : https://github.com/conan-io/conan-center-index/blob/master/.c3i/conan_v2_ready_references.yml#L699 . But it should not, because CI did not pass.
> * I couldn't get qt quick to work. Some plugins are not beeing built.
Please report a separate issue with all the details about your profile, options, command line invoked, and all the console log.

**Comment by Marwan-imverse** [comment]
Which issue should I be following to know when Qt will be V2 ready?

**Comment by ericLemanissier** [comment]
The one you will report, explaining which error you get when using qt with conan v2 :wink:

**Comment by Bobini1** [comment]
Hi, just wanted to report that I also have the same issue with `qquickabstractcolorpicker_p.h` as @Marwan-imverse. It only happens on Windows, it seems.

**Comment by ericLemanissier** [comment]
As my message above indicates, please retry with a conan local cache with a shorter path. This can be done by setting environment variable `CONAN_HOME=c:/c2` for example

**Comment by RuslanSh20** [comment]
Hi guys. Same issue.
